tests/test_rules.py

```diff
@@ -0,0 +1,322 @@
+                "removed": "",
+              },
+              "name": "processes",
+              "hostIdentifier": "hostname.local",
+              "calendarTime": "%s %s" % (now.ctime(), "UTC"),
+              "unixTime": now.strftime('%s'),
+            },
+        ]
+        node = {'host_identifier': 'hostname.local'}
+
+        rule = WhitelistRule(0, Rule.BOTH, config={
+            'field_name': 'name',
+            'whitelist': ['good', 'othergood'],
+        })
+
+        expected = RuleMatch(
+            rule_id=0,
+            action='added',
+            node=node,
+            match=data[0]['diffResults']['added'][1],
+        )
+
+        # Whitelists the two matching things, but not the middle one.
+        matches = rule.handle_log_entry({'data': data}, node)
+        assert matches == [expected]
+
+    def test_ignore_nulls(self):
+        now = dt.datetime.utcnow()
+
+        data = [
+            {
+              "diffResults": {
+                "added": [
+                  {
+                    "name": None,
+                    "path": "/usr/local/bin/good",
+                    "pid": "12345"
+                  },
+                ],
+                "removed": "",
+              },
+              "name": "processes",
+              "hostIdentifier": "hostname.local",
+              "calendarTime": "%s %s" % (now.ctime(), "UTC"),
+              "unixTime": now.strftime('%s'),
+            },
+        ]
+        node = {'host_identifier': 'hostname.local'}
+
+        rule1 = WhitelistRule(0, Rule.BOTH, config={
+            'field_name': 'name',
+            'whitelist': ['good'],
+            'ignore_null': False,
+        })
+        rule2 = WhitelistRule(1, Rule.BOTH, config={
+            'field_name': 'name',
+            'whitelist': ['good'],
+            'ignore_null': True,
+        })
+
+        expected = RuleMatch(
+            rule_id=0,
+            action='added',
+            node=node,
+            match=data[0]['diffResults']['added'][0],
+        )
+
+        assert rule1.handle_log_entry({'data': data}, node) == [expected]
+        assert rule2.handle_log_entry({'data': data}, node) == []
+
+
+class TestCountRule:
+
+    def setup_method(self, _method):
+        self.now = dt.datetime.utcnow()
+        self.fake_data = {
+            'data': [
+                {
+                    "diffResults": {
+                        "added": [],
+                        "removed": [],
+                    },
+                    "name": "fake",
+                    "hostIdentifier": "hostname.local",
+                    "calendarTime": "%s %s" % (self.now.ctime(), "UTC"),
+                    "unixTime": self.now.strftime('%s')
+                },
+            ],
+        }
+
+    def test_basic_filtering(self):
+        node = {'host_identifier': 'hostname.local'}
+
+        # Equal works
+        self.fake_data['data'][0]['diffResults']['added'] = [{'op': 'one'}, {'op': 'two'}, {'op': 'three'}]
```

Comment:
This (and the following 20 lines) are ugly as hell and I should really clean them up.